staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/policy_decision.go

```diff
@@ -29,14 +32,15 @@ const (
 )
 
 type PolicyDecision struct {
-	Kind    PolicyDecisionKind `json:"kind"`
-	Message any                `json:"message"`
+	Kind    PolicyDecisionKind  `json:"kind"`
+	Message any                 `json:"message"`
+	Reason  metav1.StatusReason `json:"reason"`
 }
 
 type PolicyDecisionWithMetadata struct {
 	PolicyDecision `json:"decision"`
-	Definition     PolicyDefinition `json:"definition"`
-	Binding        PolicyBinding    `json:"binding"`
+	Definition     *v1alpha1.ValidatingAdmissionPolicy        `json:"definition"`
+	Binding        *v1alpha1.ValidatingAdmissionPolicyBinding `json:"binding"`
 }
 
 type PolicyError struct {
```

Comment:
I believe `PolicyError` needs to be either modified or replaced such that the error returned by admission.Interface#Validate is a apierrors.StatusError including the reason and code for that reason. This will be recognized by the admission chain and used to send the appropriate HTTP response code to the client.

Also we should avoid serializing a struct to json since that results in hard to read CEL expressions like "object.spec.replicas \u003c= params.maxReplicas" which is encoded strangely because of the nesting in JSON.

https://github.com/kubernetes/kubernetes/blob/1a41cb898547683b73a6f91b6d731d439c84f48a/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L35

Reply:
Note that `FailurePolicy:Fail` related errors (compilation and mis-configuration) _might_ need to be reported using `apierrors.StatusError` to get the desired response code.

Reply:
Also note that we are only expected to return the 1st error encountered for v1alpha1, so only the decisions[0] error needs to be included in the response along with reason and code. We should priortize making that error as clean and readable as possible.

Reply:
I currently updated to still return []PolicyDecision back for each expression and handle the final err in admission code in case we want to further polish the err in the future. But still has question how to get the code